image-loader/app/controllers/ImageLoaderController.scala

```diff
@@ -185,73 +330,145 @@ class ImageLoaderController(auth: Authentication,
+      maybeImportResult =>
+        digestedFileFuture.map(digestedFile =>
+          digestedFile -> maybeImportResult
+        )
+    }.flatMap { case (digestedFile, triedStatusUri) =>
+      // convert any exception from the upload to a failure response, or pass the uploadStatusUri if successful
+      val failureResponseOrUploadStatusUri = triedStatusUri match {
+        case Failure(e: UnsupportedMimeTypeException) => Left(FailureResponse.unsupportedMimeType(e, config.supportedMimeTypes))
+        case Failure(_: IllegalArgumentException) => Left(FailureResponse.invalidUri)
+        case Failure(e: UserImageLoaderException) => Left(FailureResponse.badUserInput(e))
+        case Failure(NonFatal(_)) => Left(FailureResponse.failedUriDownload)
+        case Success(uploadStatusUri) => Right(uploadStatusUri)
       }
-
-      /* combine the import result and digest file together into a single future
-       * note that we use transformWith instead of zip here as we are still interested in value of
-       * digestedFile even if the import fails */
-      val fileAndMaybeResult: Future[(DigestedFile, Try[Result])] = importResult.transformWith{ result =>
-        digestedFileFuture.map(file => file -> result)
+      // build a Failed StatusType from the failure response or Completed if successful
+      val status = failureResponseOrUploadStatusUri match {
+        case Left(Response(_, response)) => UploadStatus(StatusType.Failed, Some(s"${response.errorKey}: ${response.errorMessage}"))
+        case Right(_) => UploadStatus(StatusType.Completed, None)
       }
-
-      // this is an unusual way of generating a result due to the need to put the error message both in the upload
-      // status table and also provide it in the response to the client.
-      fileAndMaybeResult.flatMap { case (digestedFile, importResult) =>
-        // convert exceptions to failure responses
-        val res = importResult match {
-          case Failure(e: UnsupportedMimeTypeException) => Left(FailureResponse.unsupportedMimeType(e, config.supportedMimeTypes))
-          case Failure(_: IllegalArgumentException) => Left(FailureResponse.invalidUri)
-          case Failure(e: UserImageLoaderException) => Left(FailureResponse.badUserInput(e))
-          case Failure(NonFatal(_)) => Left(FailureResponse.failedUriDownload)
-          case Success(result) => Right(result)
-        }
-        // update the upload status with the error or completion
-        val status = res match {
-          case Left(Response(_, response)) => UploadStatus(StatusType.Failed, Some(s"${response.errorKey}: ${response.errorMessage}"))
-          case Right(_) => UploadStatus(StatusType.Completed, None)
+      // try to update uploadStatusTable, log the outcome
+      uploadStatusTable.updateStatus(digestedFile.digest, status).flatMap{ outcomeOfUpdateStatus => //FIXME use set status to avoid potential ConditionNotMet (when status table rows have expired/TTL)
```

Comment:
same as earlier

Reply:
this one is the one that fires 40-50k times a day in production, since it was introduced a few years back in https://github.com/guardian/grid/pull/3126 .

Need to speak to BBC about how their process works and I'm slightly reluctant to address it in this PR, but have left the FIXME for future travellers (hopefully us / BBC soon).